htlcswitch/circuit_map.go

```diff
@@ -0,0 +1,848 @@
+// circuit key.
+func (cm *circuitMap) DeleteCircuits(inKeys ...CircuitKey) error {
+
+	var (
+		closingCircuits = make(map[CircuitKey]struct{})
+		removedCircuits = make(map[CircuitKey]*PaymentCircuit)
+	)
+
+	cm.mtx.Lock()
+	// First check that all provided keys are still known to the circuit
+	// map.
+	for _, inKey := range inKeys {
+		if _, ok := cm.pending[inKey]; !ok {
+			cm.mtx.Unlock()
+			return ErrUnknownCircuit
+		}
+	}
+
+	// If no offenders were found, remove any references to the circuit from
+	// memory, keeping track of which circuits were removed, and which ones
+	// had been marked closed. This can be used to restore these entries
+	// later if the persistent removal fails.
+	for _, inKey := range inKeys {
+		circuit := cm.pending[inKey]
+
+		delete(cm.pending, inKey)
+
+		if _, ok := cm.closed[inKey]; ok {
+			closingCircuits[inKey] = struct{}{}
+			delete(cm.closed, inKey)
+		}
+
+		if circuit.HasKeystone() {
+			delete(cm.opened, circuit.OutKey())
+			cm.removeCircuitFromHashIndex(circuit)
+		}
+
+		removedCircuits[inKey] = circuit
+	}
+	cm.mtx.Unlock()
+
+	err := cm.db.Batch(func(tx *bolt.Tx) error {
+		for _, circuit := range removedCircuits {
+			// If this htlc made it to an outgoing link, load the
+			// keystone bucket from which we will remove the
+			// outgoing circuit key.
+			if circuit.HasKeystone() {
+				keystoneBkt := tx.Bucket(circuitKeystoneKey)
+				if keystoneBkt == nil {
+					return ErrCorruptedCircuitMap
+				}
+
+				outKey := circuit.OutKey()
+
+				err := keystoneBkt.Delete(outKey.Bytes())
+				if err != nil {
+					return err
+				}
+			}
+
+			// Remove the circuit itself based on the incoming
+			// circuit key.
+			circuitBkt := tx.Bucket(circuitAddKey)
+			if circuitBkt == nil {
+				return ErrCorruptedCircuitMap
+			}
+
+			inKey := circuit.InKey()
+			if err := circuitBkt.Delete(inKey.Bytes()); err != nil {
+				return err
+			}
+		}
+
+		return nil
+	})
+
+	// Return if the write succeeded.
+	if err == nil {
+		return nil
+	}
+
+	// If the persistent changes failed, restore the circuit map to it's
+	// previous state.
+	cm.mtx.Lock()
+	for inKey, circuit := range removedCircuits {
+		cm.pending[inKey] = circuit
+
+		if _, ok := closingCircuits[inKey]; ok {
+			cm.closed[inKey] = struct{}{}
+		}
+
+		if circuit.HasKeystone() {
+			cm.opened[circuit.OutKey()] = circuit
+			cm.addCircuitToHashIndex(circuit)
+		}
+	}
+	cm.mtx.Unlock()
+
+	return nil
+}
+
+// removeCircuitFromHashIndex removes the given circuit from the hash index,
+// pruning any unnecessary memory optimistically.
+func (cm *circuitMap) removeCircuitFromHashIndex(c *PaymentCircuit) {
+	// Locate bucket containing this circuit's payment hashes.
+	circuitsWithHash, ok := cm.hashIndex[c.PaymentHash]
+	if !ok {
+		return
+	}
+
+	// Check that this circuit is a member of the set of circuit's with this
+	// payment hash.
+	outKey := c.OutKey()
+	if _, ok = circuitsWithHash[outKey]; !ok {
```

Comment:
redundant check, the following logic will just be noops if this is `!ok`

Reply:
Good point, fixing